### PR TITLE
Grandpa Pallet Pruning

### LINKED ISSUE #419: Prune Headers in Substrate Pallet
Splitting out of #369.

For storage reasons we don't want to be keeping old headers around. We should have a way to prune these headers from storage which is configurable (e.g choose how many headers you want to keep at any point in time, or how many to remove in one go). While pruning headers on the non-canonical chain is important, we should also be able to prune old headers on the canonical chain. Again, this is for storage reasons.

### PR DESCRIPTION
**Pull request by tomusdrw**
Resolves #419 

The dumbest pruning implementation, where we simply store `X` last headers around.
The implementation maintains a map of `Map<InsertionIndex, HeaderHash>` and there is a `Needle` which simply traverses all indexes from range `[0..HeadersToKeep)` and wraps around at the end. So the `ImportedHashes` structure has constant size and every time we overwrite an existing header we remove it from `ImportedHeaders` as well.

### CODE REVIEW

**Review by HCastano — APPROVED** [review]
Simple is fine here :)

File: modules/grandpa/src/lib.rs [hunk 1]
```diff
@@ -248,6 +265,15 @@ pub mod pallet {
...
 	pub(super) type BestFinalized<T: Config<I>, I: 'static = ()> = StorageValue<_, BridgedBlockHash<T, I>, ValueQuery>;
 
+	/// A ring buffer of imported hashes. Ordered by the insertion time.
+	#[pallet::storage]
+	pub(super) type ImportedHashes<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Identity /* TODO */, u32, BridgedBlockHash<T, I>>;
+
+	/// Current ring buffer position.
+	#[pallet::storage]
+	pub(super) type ImportedHashesNeedle<T: Config<I>, I: 'static = ()> = StorageValue<_, u32, ValueQuery>;
```
Comment (adoerr) [inline modules/grandpa/src/lib.rs:275]: I would prefer to use the term `index`, `position` or `pointer` here. `Needle`, AFAIK, is not commonly used.

**Review by svyatonik — APPROVED** [review]
I'm not sure, but probably benchmarks must also be updated after this @HCastano ? So that we'll always have pruning cost included into tx cost - probably it's already there? At least weights must be updated (in follow up PR)

File: modules/grandpa/src/lib.rs [hunk 2]
```diff
@@ -248,6 +265,15 @@ pub mod pallet {
 	#[pallet::storage]
 	pub(super) type BestFinalized<T: Config<I>, I: 'static = ()> = StorageValue<_, BridgedBlockHash<T, I>, ValueQuery>;
 
+	/// A ring buffer of imported hashes. Ordered by the insertion time.
+	#[pallet::storage]
+	pub(super) type ImportedHashes<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Identity /* TODO */, u32, BridgedBlockHash<T, I>>;
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply modules/grandpa/src/lib.rs:271]: Thanks for spotting that. I think it should be fine too, we could aim for some hash function that would get us perfect distribution, but it's probably irrelevant, especially that it is never expected to be part of any storage proof and performance-wise it should be negligible. We can have auditors confirm that.

**Comment by HCastano** [comment]
> I'm not sure, but probably benchmarks must also be updated after this @HCastano ? So that we'll always have pruning cost included into tx cost - probably it's already there? At least weights must be updated (in follow up PR)

Yes, they should be updated. @tomusdrw feel free to update them in a different PR since I know they can take a while to generate atm

**Pull request merged**